Source/WebCore/page/ElementTargeting.cpp

```diff
@@ -323,4 +326,45 @@ Vector<TargetedElementInfo> findTargetedElements(Page& page, TargetedElementRequ
     return results;
 }
 
+static bool setNeedsVisibilityAdjustmentRecursive(Element& element)
+{
+    if (element.isVisibilityAdjustmentRoot())
+        return false;
+
+    element.setIsVisibilityAdjustmentRoot();
+
+    CheckedPtr renderer = element.renderer();
+    if (!renderer)
+        return true;
+
+    for (auto& frameRenderer : descendantsOfType<RenderIFrame>(*renderer)) {
+        RefPtr childView = frameRenderer.childView();
+        if (!childView)
+            continue;
+
+        CheckedPtr childRenderView = childView->checkedRenderView();
+        if (!childRenderView)
+            continue;
+
+        RefPtr childDocumentElement = childRenderView->document().documentElement();
+        if (!childDocumentElement)
+            continue;
+
+        setNeedsVisibilityAdjustmentRecursive(*childDocumentElement);
+    }
+
+    element.invalidateStyleAndRenderersForSubtree();
+    return true;
+}
+
+bool adjustVisibilityForTargetedElements(const Vector<Ref<Element>>& elements)
+{
+    bool changed = false;
+    for (auto& element : elements) {
+        if (setNeedsVisibilityAdjustmentRecursive(element))
+            changed = true;
+    }
+    return changed;
```

Comment:
Ah, I think that's actually not _quite_ what I want (since I want to continue calling `setNeedsVisibilityAdjustmentRecursive` on the full list of elements, even if `changed` already becomes true). I'll leave this as-is (unless there's another more elegant way to write this?)

Reply:
ah got it; sounds good!